### DISCUSSION TITLE
Stellarium on MacOS: control Alpaca telescope?

### DISCUSSION THREAD

**Discussion by ihabmmali** [Stellarium/stellarium#3823] [Q&A]
Hello,
I'm trying to figure out how I can connect  a Stellarium instance running on MacOS to a telescope connected to a Windows mini PC running the ASCOM remote. The telescope is exposed to the network through the Alpaca driver from ASCOM remote and I can access it in other software from the same MacOS computer.

**Comment by gzotti** [comment]
It seems none of us 3 developers has time/such instruments to actually go out observing. What we know is in the User Guide. If you find out a solution, we could add a page to the User Guide for the benefit of others. As mentioned numerous times, a Qt5-based build may work better for the TelescopeControl plugin.

**Comment by ihabmmali** [comment]
thank you but  ASCOM doesn't appear to be an option in MacOS version of stellariaum...I understand that ASCOM platform is generally limited to windows which is probably why the option doesn't show up on MacOS (i.e. ascom support is not implemented outside the windows version).  There's another indirect option "remote" which seems to allow network connections but the documentation is little bit vague: "remote A device is connected to a remote computer and the software that drives it can ’talk’ to Stellarium over the network; this software can be either one of Stellarium’s stand-alone telescope servers, or a third party application."

Is there any documentation on the protocol used for this connection option?

**Reply by gzotti** [reply]
We inherited this bit of documentation from the original developers of this part. What we know is in the User Guide. What others may have found out is unknown to us.

**Comment by alex-w** [comment]
Stellarium doesn’t support Alpaca

**Comment by vyskocil** [comment]
I just wrote a Python script that bridge the old Stellarium telescope control protocol and Alpaca. I tested it on my Mac : 

https://github.com/vyskocil/Polaris/blob/main/stellarium_alpaca.py

Hope this will help !

**Reply by gzotti** [reply]
Thanks for the information. 
Suggestion:  You could parse config.ini to read long/lat information, or even get that info from a running program via RemoteControl plugin.  
I don't see Mac specifics. Should this run also on Linux?

**Reply by vyskocil** [reply]
I just removed the lat / lon parameters as it was not needed.
This script should run on linux.

**Reply by 10110111** [reply]
What is the relation between ASCOM and Alpaca? Could this script (or an implementation of a similar functionality in Stellarium itself) be a replacement for the COM-based ASCOM interface for all the ASCOM telescopes that are currently controlled from Stellarium on Windows?

**Reply by vyskocil** [reply]
Alpaca is a client server REST API which transport the ASCOM protocol (https://ascom-standards.org/About/Overview.htm)
I don't know the COM based ASCOM interface in Stellarium, maybe this has more functionalities than the old Stellarium control protocol, for which I only could implement skew telescope but abort skew is doing nothing, it is unidirectional from Stellarium -> telescope,...

**Reply by gzotti** [reply]
ASCOM is Windows-only (COM API) and the de-facto standard for astro gear on this platform, while Alpaca (AFAICR) can link *ix clients to ASCOM servers. Details: link above. 

ASCOM support was the longest-open serious wishlist entry so far (13 years on sourceforge forum?). Its implementation by an external contributor (2017?) was celebrated, but now causes crashes with Qt6. The Telescope plugin is in deep need of reorganisation by someone in the know who would collect traces of the "old telescope protocol" which apparently is still relevant but was never in our github repo or documented properly in any User Guide (history is getting lost!), collect those bits&pieces in the github repo (external telescope servers may be in separate repos of course, but should at least be available under the same user), and finally do the upgrade to system-installed unmodified INDI2, and also enhance all these protocols to AltAz mode (where available in current specs of LX200/NexStar/RTS2/ASCOM/INDI/...) which apparently has become more popular in the last 15 years (or could drive other and/or self-made devices like Arduino/RaspberryPi(co) driven stuff). I think the calibration function also is only partially implemented.  And don't forget the emergency stop! Maybe add (document in SUG) a simple tinkering project that connects a physical "red button" device via USB. While there are telescope simulators available, I think the prospective developer(s) should have at least one GOTO model compatible with our LX200 OR NexStar native protocol implementations, ASCOM and INDI, to bring the required motivation to "make it work better" for him-/herself. I think we can offer a reward for this popular wish. 

I just wonder: should we just build our Qt6-based releases without ASCOM?

**Reply by vyskocil** [reply]
I found information about the old telescope protocol in a forum and added the code in my python code to manage the link between Stellarium and a Benro Polaris motorized head (which only only a smartphone App so I had to reverse engineer its protocol).
Now there is a kickstart project (not from me) which will bring Alpaca to the Benro Polaris: https://www.kickstarter.com/projects/jdmorriso/alpaca-benro-polaris-driver
Alpaca seems to be the future of ASCOM

**Reply by 10110111** [reply]
> Alpaca seems to be the future of ASCOM

It does sound like this, but I'm just wondering if there's an already available driver/translator that would allow all Alpaca-enabled clients to control all (including legacy) ASCOM COM-based telescopes. If it does exist, then we can simply ditch the COM-based solution and implement Alpaca in Stellarium.

**Reply by vyskocil** [reply]
There is a Alpaca - ASCOM bridge software for Windows that allows to use legacy Windows only ASCOM drivers and provide Alpaca protocol that allows to remote control the device from any platform. A Mini PC with windows is then required to make the bridge until Alcapa enabled devices are appearing...

**Reply by gzotti** [reply]
And if you run this Bridge software on the same Windows PC as Stellarium, we don't need to link to the ASCOM libraries, and connection is just via network interface? This might also save us from further gray hair...

**Reply by vyskocil** [reply]
This is what I understand, windows and ASCOM users should only need to install and run the Alpaca-ASCOM bridge on the PC.

**Reply by ihabmmali** [reply]
@vyskocil thank you so much! just so I understand: your script acts like an Alpaca client to connect to server running on the PC controlling the mount and you connect to the script using the "external software or remote computer" feature in "telescopes" plugin of Stellarrium?

**Reply by vyskocil** [reply]
Yes exactly!

**Reply by ihabmmali** [reply]
Amazing! Just looking at the code, this appears to implement Stellarium to mount goto function only, it doesn't update the mount current location in Stellarium from Alpaca does it?

**Reply by vyskocil** [reply]
Yes, I don't know if the other way is possible using the "external software or remote computer" type of remote telepscope...

**Reply by ihabmmali** [reply]
Nevertheless this is really great, thank you so much! You mentioned that you found documentation around the "external software or remote computer" protocol, do you have a link to this documentation? With your base code maybe I can try to figure out the update Stellarium feature and add a function to your script (if i't seven possible). At the moment I can get target information from Stellarium into NINA using the built in server but I was originally asking about Alpaca to be able to also see where the telescope is pointing from a remote PC instead of running Stellarium locally on the mount control PC (which is not as powerful and running other things like NINA and the guiding software)

**Reply by vyskocil** [reply]
Here it is : https://free-astro.org/images/b/b7/Stellarium_telescope_protocol.txt
I missed that point the two ways are indeed documented ! I'll try to add the other way in my code...

**Reply by 10110111** [reply]
Just checked, the protocol as described in the above linked document seems to be complete enough, matching the implementation ([position sending](https://github.com/Stellarium/stellarium/blob/d81dc49aa42d0b3235987cfa4e33b87084ae8df5/plugins/TelescopeControl/src/common/Connection.cpp#L307), [data reception](https://github.com/Stellarium/stellarium/blob/d81dc49aa42d0b3235987cfa4e33b87084ae8df5/plugins/TelescopeControl/src/common/Connection.cpp#L230)). I guess this should be added to the User Guide.

Hmm, actually, this appears to have been duplicated [in TelescopeClient.cpp](https://github.com/Stellarium/stellarium/blob/d81dc49aa42d0b3235987cfa4e33b87084ae8df5/plugins/TelescopeControl/src/TelescopeClient.cpp#L264), but looks like the protocol is still the same.

**Reply by gzotti** [reply]
For those developers who don't want to look into our sources, we can of course print that ~1 page in the Guide. Is it meaningful? Or should we just add an FAQ link to the sources. Or maybe this should go into the README inside the TC plugin sources.

**Reply by 10110111** [reply]
As you can see above, I have tried looking into our sources, and I'm already puzzled by two copies of the implementation, and I have only a guess as to which one is actually used. Better have the protocol documented properly in an obvious first place to look: the User Guide. Though, if you insist, the README would be OK too—at least available from the repo.

**Reply by gzotti** [reply]
No, OK, User Guide. I have seen this protocol years ago on sf, but as the telescope servers were never inside our repo (and were "old and outdated" when I joined in 2010), I lost track. If the servers are still useful (and this is a good example), they should be kept accessible. If the server sources are lost (or even, just because it can be done and Python is so widely used), I would like to see a modern reference implementation in Python as part of github/Stellarium/telescope-server.

**Reply by gzotti** [reply]
@10110111 , @alex-w shall we already disable ASCOM for Qt6-based builds? Issues persist. Whoever has time, ASCOM7 RC1 is out with the latest Alpaca, and it seems moving to TCP/IP based Alpaca is the most meaningful solution.

**Reply by 10110111** [reply]
I think the option should just be grayed out for now, with a tooltip or parentheses telling to use a Qt5-based build. After Alpaca support is implemented, it would replace this option (possibly leaving it in Qt5-based builds for a few releases).
